File: c++/cote/GJM/간선_구분.cpp
#include <iostream>
#include <vector>
#include <random>
#define N 10
#define TEST_CASES 10

using namespace std;

vector<vector<int>> initialize(default_random_engine& gen) {
	int n = gen() % (N - 3) + 3;
	vector<vector<int>> ret(n, vector<int>());

	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n ; j++) {
			if (i == j && gen() % 1000 < 9990) {
				continue;
			}
			ret[i].push_back(j);
		}
	}

	return ret;
}

void dfs(int cur, const vector<vector<int>>& adj, vector<int>& discovered, vector<bool>& finished, int& counter) {
	discovered[cur] = counter++;

	for (const auto& next : adj[cur]) {
		cout << "(" << cur << "," << next << ") : ";
		if (discovered[next] == -1){
			cout << "tree edge" << endl;
			dfs(next, adj, discovered, finished, counter);
		} else if (discovered[cur] < discovered[next]) {
			cout << "forward edge" << endl;
		} else if (!finished[next]) {
			cout << "back edge" << endl;
 		} else {
 			cout << "cross edge" << endl;
 		}
	}
	finished[cur] = true;
}

void solve(vector<vector<int>>& adj) {
	vector<int> discovered(adj.size(), -1);
	vector<bool> finished(adj.size(), false);

	int counter = 0;

	for (int i = 0; i < adj.size(); i++) {
		if (discovered[i] == -1) {
			dfs(i, adj, discovered, finished, counter);
		}
	}
}

int main(void){
	default_random_engine gen(43);

	for (int it = 0; it < TEST_CASES; it++){
		vector<vector<int>> adj = initialize(gen);
		solve(adj);
	}

	return 0;
}

File: go/cote/GJM/간선_구분.go
package main

import(
	"math/rand"
	"slices"
	"fmt"
	"math"
)

const (
	TEST_CASES = 10
	N = 10
)

func _dfs(cur int, adj [][]int, visited []bool) {
	visited[cur] = true

	for _, next := range adj[cur] {
		if !visited[next] {
			_dfs(next, adj, visited)
		}
	}
}

func check_reachable(adj [][]int) bool {
	visited := make([]bool, len(adj))
	_dfs(0, adj, visited)

	return !slices.Contains(visited, false)
}

func initialize(random *rand.Rand) [][]int {
	n := random.Intn(N - 3) + 3
	var ret [][]int

	for {
		ret = make([][]int, n)
		for i, _ := range ret {
			for j := 0; j < n; j++ {
				if i == j || random.Float32() < 0.8  {
					continue
				}
				ret[i] = append(ret[i], j)
			}
		}

		if check_reachable(ret) {
			break
		}
	}

	return ret
}

func dfs(cur int, discovered []int, finished []bool, counter *int, adj [][]int) int {

	discovered[cur] = *counter
	(*counter)++

	ret := math.MaxInt32

	for _, next := range adj[cur] {

		if discovered[next] == -1 {
			fmt.Printf("tree edge(%d,%d) ",cur, next)
			dfs(next, discovered, finished, counter, adj)
		
		} else if discovered[next] > discovered[cur] {
			fmt.Printf("forward edge(%d,%d) ",cur, next)

		} else if !finished[next] {
			fmt.Printf("backward edge(%d,%d) ",cur, next)

		} else {
			fmt.Printf("cross edge(%d,%d) ",cur, next)
		}
	}

	finished[cur] = true

	return ret
}


func solve(adj [][]int) {
	discovered := make([]int, len(adj))	
	finished := make([]bool, len(adj))
	counter := 0

	for i, _ := range discovered {
		discovered[i] = -1
	}

	for i := 0; i < len(adj); i++ {
		if discovered[i] == -1 {
			dfs(i, discovered, finished, &counter, adj)
		}
	}
}

func main() {
	random := rand.New(rand.NewSource(43))

	for it := 0; it < TEST_CASES; it++ {
		adj := initialize(random)
		solve(adj)
		fmt.Println()
	}
}